README.md
# google_maps_routes

This is a package for creating routes on [google maps](https://pub.dev/packages/google_maps_flutter) using the directions API from Google.
Routes are drawn with help of the [flutter_polyline_points](https://pub.dev/packages/flutter_polyline_points) package.

## Screenshot
<img src="https://raw.githubusercontent.com/aikenahac/google_maps_routes/master/assets/example-route.jpeg" alt="example-route" width="200"/>

## Usage

To use the google_maps_routes package, first [depend on it](https://flutter.dev/docs/development/packages-and-plugins/using-packages).

Then import it where you want to use it

```dart
import 'package:google_maps_routes/google_maps_routes.dart';
```

To use it you need a `List` of points like this:

```dart
/// LatLng is included in google_maps_flutter
List<LatLng> points = [
    LatLng(latitude, longitude),
    LatLng(latitude, longitude),
    LatLng(latitude, longitude),
    LatLng(latitude, longitude),
];
```

Then you need to instantiate this plugin

```dart
MapsRoutes route = new MapsRoutes();
```

To create a route, just call the following function:

```dart
/// routeName is a string and can be anything; it can include spaces but
/// they will be replaced with a dash
/// example: 'Example route' turns into 'Example-route'

/// color is a Color type and will be used as the polyline color:
/// example: Color.fromRGBO(130, 78, 210, 1.0)

/// googleApyKey is a string and is a google directions API key
/// example: get it at
/// https://developers.google.com/maps/documentation/directions/get-api-key
await route.drawRoute(points, routeName, color, googleApiKey);
```

To use a different travel mode, use the travelMode parameter.

```dart
/// Options: driving, bicycling, transit, walking
await route.drawRoute(
    points,
    routeName,
    color,
    googleApiKey,
    travelMode: TravelModes.walking
);
```

To display the polylines you need to add a polylines parameter to your google map widget.

```dart
GoogleMap(
    polylines: route.routes
)
```

If you ever want to clear the routes, just call:

```dart
route.routes.clear();
```

If you want to calculate the distance of the route use you call the following function with the same `List` of points you used for drawing the route. You can also use it by itself.

```dart
/// Initialize a calculator
DistanceCalculator distanceCalculator = new DistanceCalculator();

/// Call a function that returns a string
/// Pass it the List of points
/// By default it will return with 1 decimal point but you can pass an 
/// option that changes that 
distanceCalculator.calculateRouteDistance(points, decimals: 1);
```

### Credits

- [google maps](https://pub.dev/packages/google_maps_flutter) for providing the map
- [flutter_polyline_points](https://pub.dev/packages/flutter_polyline_points) for the usage of directions api


CHANGELOG.md
## [0.0.1] - Create package

* Draw routes on google maps with the directions API

## [0.0.2] - Provide better documentation

## [1.0.0] - Add support for travel modes

## [1.0.1] - Update readme

## [1.1.0] - Add option for distance calculation

## [1.1.1] - Provide better code documentation

## [1.1.2] - Bug fixes

lib/src/route_renderer.dart
/// Importing necessary packages
import 'package:google_maps_flutter/google_maps_flutter.dart';
import 'package:flutter_polyline_points/flutter_polyline_points.dart';
import 'package:flutter/material.dart';

/// Choices between modes of travel
enum TravelModes { driving, bicycling, transit, walking }

/// A library for creating a route over google maps with the directions API
class MapsRoutes {
  Set<Polyline> routes = {};

  /// Replaces whitespaces in string with a single dash
  String _replaceWhiteSpaces(String str) {
    return str.replaceAll(' ', '-');
  }

  /// Function that creates a route between two points with directions API
  _createRouteFragment(
      double? startLat,
      double? startLon,
      double? endLat,
      double? endLon,
      String routeName,
      Color routeColor,
      String googleApiKey,
      TravelMode travelMode) async {
    late PolylinePoints routePoints = PolylinePoints();
    List<LatLng> routeCoordinates = [];

    /// If the coordinates are null, it returns an empty route
    if (startLat == null ||
        startLon == null ||
        endLat == null ||
        endLon == null) {
      return;
    }

    /// If the coordinates are not null, it creates a route between the two points
    PolylineResult result = await routePoints.getRouteBetweenCoordinates(
        googleApiKey,
        PointLatLng(startLat, startLon),
        PointLatLng(endLat, endLon),
        travelMode: travelMode);

    /// Adds coordinates to the route coordinates list
    if (result.points.isNotEmpty) {
      result.points.forEach((PointLatLng point) {
        routeCoordinates.add(LatLng(point.latitude, point.longitude));
      });
    }

    /// Sets a polyline ID
    PolylineId id = PolylineId('route-$routeName');

    /// Creates a polyline with the route coordinates
    Polyline route = Polyline(
        polylineId: id, color: routeColor, points: routeCoordinates, width: 3);

    /// Adds the route to the routes list
    routes.add(route);
  }

  /// Function that creates the actual route between multiple points
  Future<void> drawRoute(List<LatLng> points, String routeName,
      Color routeColor, String googleApiKey,
      {TravelModes? travelMode}) async {
    var previousPoint;
    TravelMode travelType;

    /// Checks which travel mode is defined in the parameters
    if (travelMode != null) {
      switch (travelMode) {
        case TravelModes.driving:
          travelType = TravelMode.driving;
          break;
        case TravelModes.bicycling:
          travelType = TravelMode.bicycling;
          break;
        case TravelModes.transit:
          travelType = TravelMode.transit;
          break;
        case TravelModes.walking:
          travelType = TravelMode.walking;
          break;
        default:
          travelType = TravelMode.driving;
      }
    }

    /// If the travel mode is not defined, it uses the default travel mode
    else {
      travelType = TravelMode.driving;
    }

    /// Iterates over the points and creates a route between each pair
    for (var i = 0; i < points.length; i++) {
      var point = points[i];

      /// If the previous point is null it creates a route
      /// from the first and second point
      if (previousPoint == null) {
        var nextPoint = points[i + 1];
        previousPoint = point;
        await _createRouteFragment(
            point.latitude,
            point.longitude,
            nextPoint.latitude,
            nextPoint.longitude,
            '${_replaceWhiteSpaces(routeName)}+$i',
            routeColor,
            googleApiKey,
            travelType);
      }

      /// If the previous point is not null it creates a route
      /// between the previous and current point
      else {
        await _createRouteFragment(
            previousPoint.latitude,
            previousPoint.longitude,
            point.latitude,
            point.longitude,
            '${_replaceWhiteSpaces(routeName)}+$i',
            routeColor,
            googleApiKey,
            travelType);
      }
    }
  }
}


lib/src/distance_calculator.dart
/// Importing necessary packages
import 'package:google_maps_flutter/google_maps_flutter.dart';
import 'dart:math' show cos, sqrt, asin;

/// Class for the distance calculator
class DistanceCalculator {
  /// Calculates the distance between two points
  double _routeDistance(lat1, lon1, lat2, lon2) {
    var p = 0.017453292519943295;
    var c = cos;
    var a = 0.5 -
        c(((lat2 - lat1) as num) * p) / 2 +
        c((lat1 as num) * p) *
            c((lat2 as num) * p) *
            (1 - c(((lon2 - lon1) as num) * p)) /
            2;
    return 12742 * asin(sqrt(a));
  }

  /// Calculates the distance between all points in a list
  String calculateRouteDistance(List<LatLng> points, {decimals}) {
    double totalDistance = 0.0;

    /// Defaults the number of decimals to 1 if not specified
    decimals ??= 1;

    /// Iterates through all points in the list and calculates
    /// the distance between each point
    for (int i = 0; i < points.length - 1; i++) {
      totalDistance += _routeDistance(
        points[i].latitude,
        points[i].longitude,
        points[i + 1].latitude,
        points[i + 1].longitude,
      );
    }

    print('TOTAL DISTANCE: ${totalDistance.toStringAsFixed(decimals)} km');

    /// Returns the total distance in km
    return '${totalDistance.toStringAsFixed(decimals)} km';
  }
}


example/lib/main.dart
import 'dart:async';

import 'package:flutter/material.dart';
import 'package:google_maps_flutter/google_maps_flutter.dart';
import 'package:google_maps_routes/google_maps_routes.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Google Maps Routes Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: MapsRoutesExample(title: 'GMR Demo Home'),
    );
  }
}

class MapsRoutesExample extends StatefulWidget {
  MapsRoutesExample({Key? key, required this.title}) : super(key: key);
  final String title;

  @override
  _MapsRoutesExampleState createState() => _MapsRoutesExampleState();
}

class _MapsRoutesExampleState extends State<MapsRoutesExample> {
  Completer<GoogleMapController> _controller = Completer();

  List<LatLng> points = [
    LatLng(45.82917150748776, 14.63705454546316),
    LatLng(45.833828635680355, 14.636544256202207),
    LatLng(45.851254420031296, 14.624331708344428),
    LatLng(45.84794984187217, 14.605434384742317)
  ];

  MapsRoutes route = new MapsRoutes();
  DistanceCalculator distanceCalculator = new DistanceCalculator();
  String googleApiKey = 'YOUR KEY HERE';
  String totalDistance = 'No route';

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: Stack(
        children: [
          Align(
            alignment: Alignment.center,
            child: GoogleMap(
              zoomControlsEnabled: false,
              polylines: route.routes,
              initialCameraPosition: const CameraPosition(
                zoom: 15.0,
                target: LatLng(45.82917150748776, 14.63705454546316),
              ),
              onMapCreated: (GoogleMapController controller) {
                _controller.complete(controller);
              },
            ),
          ),
          Padding(
            padding: const EdgeInsets.all(8.0),
            child: Align(
              alignment: Alignment.bottomCenter,
              child: Container(
                  width: 200,
                  height: 50,
                  decoration: BoxDecoration(
                      color: Colors.white,
                      borderRadius: BorderRadius.circular(15.0)),
                  child: Align(
                    alignment: Alignment.center,
                    child:
                        Text(totalDistance, style: TextStyle(fontSize: 25.0)),
                  )),
            ),
          )
        ],
      ),
      floatingActionButton: FloatingActionButton(
        onPressed: () async {
          await route.drawRoute(points, 'Test routes',
              Color.fromRGBO(130, 78, 210, 1.0), googleApiKey,
              travelMode: TravelModes.walking);
          setState(() {
            totalDistance =
                distanceCalculator.calculateRouteDistance(points, decimals: 1);
          });
        },
      ),
    );
  }
}
